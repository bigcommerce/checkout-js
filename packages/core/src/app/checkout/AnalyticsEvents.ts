let boltTracker: Window["BoltTrack"] = {
  recordEvent: (event: string) => console.log("--bolt bigc--: NOOP ", event),
};

const eventLog: string[] = [];

export const AnalyticsEvents = {
  init: () => {
    try {
      console.log("--bolt bigc--: intializing Bolt-BigC analytics events");
      if (window && window.BoltTrack) {
        boltTracker = window.BoltTrack;
        console.log("--bolt bigc--: successfully assigned window BoltTrack");
        // immediately send checkout load success event
        AnalyticsEvents.emitEvent("Checkout load success");
      }
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  emitEvent: (eventName: string) => {
    try {
      console.log("--bolt bigc--: emitting analytics event ", eventName);
      const props: any = {
        nextState: eventName,
        prevState: "",
      };
      if (eventLog.length > 0) {
        props.prevState = eventLog[0];
      }
      boltTracker.recordEvent("CheckoutFunnelTransition", props);
      // add latest event to beginning of log array
      eventLog.unshift(eventName);
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  onBeforeUnload: () => {
    AnalyticsEvents.emitEvent("Exit");
  },
};
