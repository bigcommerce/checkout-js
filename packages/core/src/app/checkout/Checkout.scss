@import '../ui/Base';

$color-secondary: #757575;
$color-link: #476BEF;
$fontFamily-montserrat: 'Montserrat';

// optimizedCheckout-headingPrimary
.header, .header-secondary, .sub-header {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("large");
    font-weight: fontWeight("bold");
}

.header-secondary {
    color: $color-secondary;
}

// optimizedCheckout-headingSecondary
.sub-header {
    font-size: fontSize("smaller");
}

// optimizedCheckout-contentPrimary
.body-regular, .body-cta, .body-medium, .body-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("base");
    font-weight: fontWeight("normal");
}

.body-cta {
    color: $color-link;
}

.body-medium, .body-bold {
    font-weight: fontWeight("medium");
}

.body-bold {
    font-size: fontSize("base");
    font-weight: fontWeight("bold");
}

.sub-text, .sub-text-medium, .sub-text-bold {
    color: $color-black;
    font-family: $fontFamily-montserrat;
    font-size: fontSize("smallest");
    font-weight: fontWeight("normal");
}

.sub-text-medium {
    color: $color-secondary;
    font-weight: fontWeight("medium");
}

.sub-text-bold {
    font-weight: fontWeight("bold");
}

// overrides 
.body-bold, .body-medium {
    &.button--primary:not(:disabled){
        color: $color-white;
    }
}

.form-checkbox + .form-label,
.form-radio + .form-label {
    &.body-regular {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("normal");
    }
}

.shippingOptionLabel {
    .shippingOption-price, .shippingOption-desc {
        &.body-medium {
            color: $color-black;
            font-family: $fontFamily-montserrat;
            font-size: fontSize("base");
            font-weight: fontWeight("medium");
        }
    }
}

.body-medium {
     small {
        color: $color-black;
        font-family: $fontFamily-montserrat;
        font-size: fontSize("base");
        font-weight: fontWeight("medium");
    }
}

.add-consignment-button {
    &.body-regular {
        color: $color-black;
    }
}
