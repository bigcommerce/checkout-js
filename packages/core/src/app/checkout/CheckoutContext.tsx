import { CheckoutSelectors, CheckoutService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface CheckoutContextProps {
    checkoutService: CheckoutService;
    checkoutState: CheckoutSelectors;
}

const CheckoutContext = createContext<CheckoutContextProps | undefined>(undefined);

export default CheckoutContext;
