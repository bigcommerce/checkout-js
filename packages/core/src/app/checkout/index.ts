export { RenderCheckout, RenderCheckoutOptions } from './renderCheckout';
export { default as CheckoutContext, CheckoutContextProps } from './CheckoutContext';
export {
    default as CheckoutProvider,
    CheckoutProviderProps,
    CheckoutProviderState,
} from './CheckoutProvider';
export { default as withCheckout, WithCheckoutProps } from './withCheckout';
export { default as CheckoutSupport } from './CheckoutSupport';
export { default as NoopCheckoutSupport } from './NoopCheckoutSupport';
export { default as navigateToOrderConfirmation } from './navigateToOrderConfirmation';
