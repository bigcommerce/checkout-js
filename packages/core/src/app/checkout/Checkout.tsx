import {
    Address,
    Cart,
    CartChangedError,
    CheckoutParams,
    CheckoutSelectors,
    Consignment,
    EmbeddedCheckoutMessenger,
    EmbeddedCheckoutMessengerOptions,
    FlashMessage,
    Promotion,
    RequestOptions,
} from '@bigcommerce/checkout-sdk';
import classNames from 'classnames';
import { find, findIndex } from 'lodash';
import React, { Component, lazy, ReactNode } from 'react';

import { AnalyticsContextProps } from '@bigcommerce/checkout/analytics';
import { ExtensionContextProps, withExtension } from '@bigcommerce/checkout/checkout-extension';
import { ErrorLogger } from '@bigcommerce/checkout/error-handling-utils';
import { TranslatedString, withLanguage, WithLanguageProps } from '@bigcommerce/checkout/locale';
import { AddressFormSkeleton, ChecklistSkeleton } from '@bigcommerce/checkout/ui';

import { withAnalytics } from '../analytics';
import { StaticBillingAddress } from '../billing';
import { EmptyCartMessage } from '../cart';
import { withCheckout } from '../checkout';
import { CustomError, ErrorModal, isCustomError } from '../common/error';
import { retry } from '../common/utility';
import {
    CheckoutButtonContainer,
    CheckoutSuggestion,
    Customer,
    CustomerInfo,
    CustomerSignOutEvent,
    CustomerViewType,
} from '../customer';
import { EmbeddedCheckoutStylesheet, isEmbedded } from '../embeddedCheckout';
import { PromotionBannerList } from '../promotion';
import { hasSelectedShippingOptions, isUsingMultiShipping, StaticConsignment } from '../shipping';
import { ShippingOptionExpiredError } from '../shipping/shippingOption';
import { LazyContainer, LoadingNotification, LoadingOverlay } from '../ui/loading';
import { MobileView } from '../ui/responsive';

import CheckoutStep from './CheckoutStep';
import CheckoutStepStatus from './CheckoutStepStatus';
import CheckoutStepType from './CheckoutStepType';
import CheckoutSupport from './CheckoutSupport';
import mapToCheckoutProps from './mapToCheckoutProps';
import navigateToOrderConfirmation from './navigateToOrderConfirmation';

const Billing = lazy(() =>
    retry(
        () =>
            import(
                /* webpackChunkName: "billing" */
                '../billing/Billing'
            ),
    ),
);

const CartSummary = lazy(() =>
    retry(
        () =>
            import(
                /* webpackChunkName: "cart-summary" */
                '../cart/CartSummary'
            ),
    ),
);

const CartSummaryDrawer = lazy(() =>
    retry(
        () =>
            import(
                /* webpackChunkName: "cart-summary-drawer" */
                '../cart/CartSummaryDrawer'
            ),
    ),
);

const Payment = lazy(() =>
    retry(
        () =>
            import(
                /* webpackChunkName: "payment" */
                '../payment/Payment'
            ),
    ),
);

const Shipping = lazy(() =>
    retry(
        () =>
            import(
                /* webpackChunkName: "shipping" */
                '../shipping/Shipping'
            ),
    ),
);

export interface CheckoutProps {
    checkoutId: string;
    containerId: string;
    embeddedStylesheet: EmbeddedCheckoutStylesheet;
    embeddedSupport: CheckoutSupport;
    errorLogger: ErrorLogger;
    createEmbeddedMessenger(options: EmbeddedCheckoutMessengerOptions): EmbeddedCheckoutMessenger;
}

export interface CheckoutState {
    activeStepType?: CheckoutStepType;
    isBillingSameAsShipping: boolean;
    customerViewType?: CustomerViewType;
    defaultStepType?: CheckoutStepType;
    error?: Error;
    flashMessages?: FlashMessage[];
    isMultiShippingMode: boolean;
    isCartEmpty: boolean;
    isRedirecting: boolean;
    hasSelectedShippingOptions: boolean;
    isHidingStepNumbers: boolean;
    isSubscribed: boolean;
}

export interface WithCheckoutProps {
    billingAddress?: Address;
    cart?: Cart;
    consignments?: Consignment[];
    error?: Error;
    hasCartChanged: boolean;
    flashMessages?: FlashMessage[];
    isGuestEnabled: boolean;
    isLoadingCheckout: boolean;
    isPending: boolean;
    isPriceHiddenFromGuests: boolean;
    isShowingWalletButtonsOnTop: boolean;
    loginUrl: string;
    cartUrl: string;
    createAccountUrl: string;
    canCreateAccountInCheckout: boolean;
    promotions?: Promotion[];
    steps: CheckoutStepStatus[];
    clearError(error?: Error): void;
    loadCheckout(id: string, options?: RequestOptions<CheckoutParams>): Promise<CheckoutSelectors>;
    subscribeToConsignments(subscriber: (state: CheckoutSelectors) => void): () => void;
}

class Checkout extends Component<
    CheckoutProps &
        WithCheckoutProps &
        WithLanguageProps &
        AnalyticsContextProps &
        ExtensionContextProps,
    CheckoutState
> {
    state: CheckoutState = {
        isBillingSameAsShipping: true,
        isCartEmpty: false,
        isRedirecting: false,
        isMultiShippingMode: false,
        hasSelectedShippingOptions: false,
        isHidingStepNumbers: true,
        isSubscribed: false,
    };

    private embeddedMessenger?: EmbeddedCheckoutMessenger;
    private unsubscribeFromConsignments?: () => void;

    componentWillUnmount(): void {
        if (this.unsubscribeFromConsignments) {
            this.unsubscribeFromConsignments();
            this.unsubscribeFromConsignments = undefined;
        }

        window.removeEventListener('beforeunload', this.handleBeforeExit);
        this.handleBeforeExit();
    }

    async componentDidMount(): Promise<void> {
        const {
            analyticsTracker,
            checkoutId,
            containerId,
            createEmbeddedMessenger,
            embeddedStylesheet,
            extensionService,
            loadCheckout,
            subscribeToConsignments,
            isExtensionEnabled,
        } = this.props;

        try {
            const { data } = await loadCheckout(checkoutId, {
                params: {
                    include: [
                        'cart.lineItems.physicalItems.categoryNames',
                        'cart.lineItems.digitalItems.categoryNames',
                    ] as any, // FIXME: Currently the enum is not exported so it can't be used here.
                },
            });
            const { links: { siteLink = '' } = {} } = data.getConfig() || {};
            const errorFlashMessages = data.getFlashMessages('error') || [];

            if (errorFlashMessages.length) {
                const { language } = this.props;

                this.setState({
                    error: new CustomError({
                        title:
                            errorFlashMessages[0].title ||
                            language.translate('common.error_heading'),
                        message: errorFlashMessages[0].message,
                        data: {},
                        name: 'default',
                    }),
                });
            }

            const messenger = createEmbeddedMessenger({ parentOrigin: siteLink });

            this.unsubscribeFromConsignments = subscribeToConsignments(
                this.handleConsignmentsUpdated,
            );
            this.embeddedMessenger = messenger;
            messenger.receiveStyles((styles) => embeddedStylesheet.append(styles));
            messenger.postFrameLoaded({ contentId: containerId });
            messenger.postLoaded();

            analyticsTracker.checkoutBegin();

            const consignments = data.getConsignments();
            const cart = data.getCart();

            const hasMultiShippingEnabled =
                data.getConfig()?.checkoutSettings.hasMultiShippingEnabled;
            const checkoutBillingSameAsShippingEnabled =
                data.getConfig()?.checkoutSettings.checkoutBillingSameAsShippingEnabled ?? true;
            const removeStepNumbersFlag =
              data.getConfig()?.checkoutSettings.features['CHECKOUT-7255.remove_checkout_step_numbers'] ??
              false;
            const defaultNewsletterSignupOption =
                data.getConfig()?.shopperConfig.defaultNewsletterSignup ??
                false;
            const isMultiShippingMode =
                !!cart &&
                !!consignments &&
                hasMultiShippingEnabled &&
                isUsingMultiShipping(consignments, cart.lineItems);

            this.setState({
                isBillingSameAsShipping: checkoutBillingSameAsShippingEnabled,
                isHidingStepNumbers: removeStepNumbersFlag,
                isSubscribed: defaultNewsletterSignupOption,
            });

            if (isMultiShippingMode) {
                this.setState({ isMultiShippingMode }, this.handleReady);
            } else {
                this.handleReady();
            }

            window.addEventListener('beforeunload', this.handleBeforeExit);

            if (isExtensionEnabled()) {
                await extensionService.loadExtensions();
            }
        } catch (error) {
            if (error instanceof Error) {
                this.handleUnhandledError(error);
            }
        }
    }

    render(): ReactNode {
        const { error, isHidingStepNumbers } = this.state;
        let errorModal = null;

        if (error) {
            if (isCustomError(error)) {
                errorModal = (
                    <ErrorModal
                        error={error}
                        onClose={this.handleCloseErrorModal}
                        title={error.title}
                    />
                );
            } else {
                errorModal = <ErrorModal error={error} onClose={this.handleCloseErrorModal} />;
            }
        }

        return (
            <div className={classNames({ 'is-embedded': isEmbedded(), 'remove-checkout-step-numbers': isHidingStepNumbers })}>
                <div className="layout optimizedCheckout-contentPrimary">
                    {this.renderContent()}
                </div>
                {errorModal}
            </div>
        );
    }

    private renderContent(): ReactNode {
        const { isPending, loginUrl, promotions = [], steps, isShowingWalletButtonsOnTop, extensionState } = this.props;

        const { activeStepType, defaultStepType, isCartEmpty, isRedirecting } = this.state;

        if (isCartEmpty) {
            return <EmptyCartMessage loginUrl={loginUrl} waitInterval={3000} />;
        }

        const isPaymentStepActive = activeStepType
            ? activeStepType === CheckoutStepType.Payment
            : defaultStepType === CheckoutStepType.Payment;

        return (
            <LoadingOverlay hideContentWhenLoading isLoading={isRedirecting}>
                <div className="layout-main">
                    <LoadingNotification isLoading={(!isShowingWalletButtonsOnTop && isPending) || extensionState.isShowingLoadingIndicator} />

                    <PromotionBannerList promotions={promotions} />

                    {isShowingWalletButtonsOnTop && (
                        <CheckoutButtonContainer
                            checkEmbeddedSupport={this.checkEmbeddedSupport}
                            isPaymentStepActive={isPaymentStepActive}
                            onUnhandledError={this.handleUnhandledError}
                        />
                    )}

                    <ol className="checkout-steps">
                        {steps
                            .filter((step) => step.isRequired)
                            .map((step) =>
                                this.renderStep({
                                    ...step,
                                    isActive: activeStepType
                                        ? activeStepType === step.type
                                        : defaultStepType === step.type,
                                    isBusy: isPending,
                                }),
                            )}
                    </ol>
                </div>

                {this.renderCartSummary()}
            </LoadingOverlay>
        );
    }

    private renderStep(step: CheckoutStepStatus): ReactNode {
        switch (step.type) {
            case CheckoutStepType.Customer:
                return this.renderCustomerStep(step);

            case CheckoutStepType.Shipping:
                return this.renderShippingStep(step);

            case CheckoutStepType.Billing:
                return this.renderBillingStep(step);

            case CheckoutStepType.Payment:
                return this.renderPaymentStep(step);

            default:
                return null;
        }
    }

    private renderCustomerStep(step: CheckoutStepStatus): ReactNode {
        const { isGuestEnabled, isShowingWalletButtonsOnTop } = this.props;
        const {
            customerViewType = isGuestEnabled ? CustomerViewType.Guest : CustomerViewType.Login,
            isSubscribed,
        } = this.state;

        return (
            <CheckoutStep
                {...step}
                heading={<TranslatedString id="customer.customer_heading" />}
                key={step.type}
                onEdit={this.handleEditStep}
                onExpanded={this.handleExpanded}
                suggestion={<CheckoutSuggestion />}
                summary={
                    <CustomerInfo
                        onSignOut={this.handleSignOut}
                        onSignOutError={this.handleError}
                    />
                }
            >
                <Customer
                    checkEmbeddedSupport={this.checkEmbeddedSupport}
                    isEmbedded={isEmbedded()}
                    isSubscribed={isSubscribed}
                    isWalletButtonsOnTop = {isShowingWalletButtonsOnTop }
                    onAccountCreated={this.navigateToNextIncompleteStep}
                    onChangeViewType={this.setCustomerViewType}
                    onContinueAsGuest={this.navigateToNextIncompleteStep}
                    onContinueAsGuestError={this.handleError}
                    onReady={this.handleReady}
                    onSignIn={this.navigateToNextIncompleteStep}
                    onSignInError={this.handleError}
                    onSubscribeToNewsletter={this.handleNewsletterSubscription}
                    onUnhandledError={this.handleUnhandledError}
                    step={step}
                    viewType={customerViewType}
                />
            </CheckoutStep>
        );
    }

    private renderShippingStep(step: CheckoutStepStatus): ReactNode {
        const { hasCartChanged, cart, consignments = [] } = this.props;

        const { isBillingSameAsShipping, isMultiShippingMode } = this.state;

        if (!cart) {
            return;
        }

        return (
            <CheckoutStep
                {...step}
                heading={<TranslatedString id="shipping.shipping_heading" />}
                key={step.type}
                onEdit={this.handleEditStep}
                onExpanded={this.handleExpanded}
                summary={consignments.map((consignment) => (
                    <div className="staticConsignmentContainer" key={consignment.id}>
                        <StaticConsignment
                            cart={cart}
                            compactView={consignments.length < 2}
                            consignment={consignment}
                        />
                    </div>
                ))}
            >
                <LazyContainer loadingSkeleton={<AddressFormSkeleton />}>
                    <Shipping
                        cartHasChanged={hasCartChanged}
                        isBillingSameAsShipping={isBillingSameAsShipping}
                        isMultiShippingMode={isMultiShippingMode}
                        navigateNextStep={this.handleShippingNextStep}
                        onCreateAccount={this.handleShippingCreateAccount}
                        onReady={this.handleReady}
                        onSignIn={this.handleShippingSignIn}
                        onToggleMultiShipping={this.handleToggleMultiShipping}
                        onUnhandledError={this.handleUnhandledError}
                        step={step}
                    />
                </LazyContainer>
            </CheckoutStep>
        );
    }

    private renderBillingStep(step: CheckoutStepStatus): ReactNode {
        const { billingAddress } = this.props;

        return (
            <CheckoutStep
                {...step}
                heading={<TranslatedString id="billing.billing_heading" />}
                key={step.type}
                onEdit={this.handleEditStep}
                onExpanded={this.handleExpanded}
                summary={billingAddress && <StaticBillingAddress address={billingAddress} />}
            >
                <LazyContainer loadingSkeleton={<AddressFormSkeleton />}>
                    <Billing
                        navigateNextStep={this.navigateToNextIncompleteStep}
                        onReady={this.handleReady}
                        onUnhandledError={this.handleUnhandledError}
                    />
                </LazyContainer>
            </CheckoutStep>
        );
    }

    private renderPaymentStep(step: CheckoutStepStatus): ReactNode {
        const { consignments, cart, errorLogger } = this.props;

        return (
            <CheckoutStep
                {...step}
                heading={<TranslatedString id="payment.payment_heading" />}
                key={step.type}
                onEdit={this.handleEditStep}
                onExpanded={this.handleExpanded}
            >
                <LazyContainer loadingSkeleton={<ChecklistSkeleton />}>
                    <Payment
                        checkEmbeddedSupport={this.checkEmbeddedSupport}
                        errorLogger={errorLogger}
                        isEmbedded={isEmbedded()}
                        isUsingMultiShipping={
                            cart && consignments
                                ? isUsingMultiShipping(consignments, cart.lineItems)
                                : false
                        }
                        onCartChangedError={this.handleCartChangedError}
                        onFinalize={this.navigateToOrderConfirmation}
                        onReady={this.handleReady}
                        onSubmit={this.navigateToOrderConfirmation}
                        onSubmitError={this.handleError}
                        onUnhandledError={this.handleUnhandledError}
                    />
                </LazyContainer>
            </CheckoutStep>
        );
    }

    private renderCartSummary(): ReactNode {
        return (
            <MobileView>
                {(matched) => {
                    if (matched) {
                        return (
                            <LazyContainer>
                                <CartSummaryDrawer />
                            </LazyContainer>
                        );
                    }

                    return (
                        <aside className="layout-cart">
                            <LazyContainer>
                                <CartSummary />
                            </LazyContainer>
                        </aside>
                    );
                }}
            </MobileView>
        );
    }

    private navigateToStep(type: CheckoutStepType, options?: { isDefault?: boolean }): void {
        const { clearError, error, steps } = this.props;
        const { activeStepType } = this.state;
        const step = find(steps, { type });

        if (!step) {
            return;
        }

        if (activeStepType === step.type) {
            return;
        }

        if (options && options.isDefault) {
            this.setState({ defaultStepType: step.type });
        } else {
            this.setState({ activeStepType: step.type });
        }

        if (error) {
            clearError(error);
        }
    }

    private handleToggleMultiShipping: () => void = () => {
        const { isMultiShippingMode } = this.state;

        this.setState({ isMultiShippingMode: !isMultiShippingMode });
    };

    private navigateToNextIncompleteStep: (options?: { isDefault?: boolean }) => void = (
        options,
    ) => {
        const { steps, analyticsTracker } = this.props;
        const activeStepIndex = findIndex(steps, { isActive: true });
        const activeStep = activeStepIndex >= 0 && steps[activeStepIndex];

        if (!activeStep) {
            return;
        }

        const previousStep = steps[Math.max(activeStepIndex - 1, 0)];

        if (previousStep) {
            analyticsTracker.trackStepCompleted(previousStep.type);
        }

        this.navigateToStep(activeStep.type, options);
    };

    private navigateToOrderConfirmation: (orderId?: number) => void = (orderId) => {
        const { steps, analyticsTracker } = this.props;

        analyticsTracker.trackStepCompleted(steps[steps.length - 1].type);

        if (this.embeddedMessenger) {
            this.embeddedMessenger.postComplete();
        }

        this.setState({ isRedirecting: true }, () => {
            navigateToOrderConfirmation(orderId);
        });
    };

    private checkEmbeddedSupport: (methodIds: string[]) => boolean = (methodIds) => {
        const { embeddedSupport } = this.props;

        return embeddedSupport.isSupported(...methodIds);
    };

    private handleCartChangedError: (error: CartChangedError) => void = () => {
        this.navigateToStep(CheckoutStepType.Shipping);
    };

    private handleConsignmentsUpdated: (state: CheckoutSelectors) => void = ({ data }) => {
        const { hasSelectedShippingOptions: prevHasSelectedShippingOptions, activeStepType, defaultStepType } =
            this.state;

        const { steps } = this.props;

        const newHasSelectedShippingOptions = hasSelectedShippingOptions(
            data.getConsignments() || [],
        );

        const isDefaultStepPaymentOrBilling =
            !activeStepType &&
            (defaultStepType === CheckoutStepType.Payment ||
                defaultStepType === CheckoutStepType.Billing);

        const isShippingStepFinished =
            findIndex(steps, { type: CheckoutStepType.Shipping }) <
                findIndex(steps, { type: activeStepType }) || isDefaultStepPaymentOrBilling;

        if (
            prevHasSelectedShippingOptions &&
            !newHasSelectedShippingOptions &&
            isShippingStepFinished
        ) {
            this.navigateToStep(CheckoutStepType.Shipping);
            this.setState({ error: new ShippingOptionExpiredError() });
        }

        this.setState({ hasSelectedShippingOptions: newHasSelectedShippingOptions });
    };

    private handleCloseErrorModal: () => void = () => {
        this.setState({ error: undefined });
    };

    private handleExpanded: (type: CheckoutStepType) => void = (type) => {
        const { analyticsTracker } = this.props;

        analyticsTracker.trackStepViewed(type);
    };

    private handleUnhandledError: (error: Error) => void = (error) => {
        this.handleError(error);

        // For errors that are not caught and handled by child components, we
        // handle them here by displaying a generic error modal to the shopper.
        this.setState({ error });
    };

    private handleError: (error: Error) => void = (error) => {
        const { errorLogger } = this.props;

        errorLogger.log(error);

        if (this.embeddedMessenger) {
            this.embeddedMessenger.postError(error);
        }
    };

    private handleEditStep: (type: CheckoutStepType) => void = (type) => {
        this.navigateToStep(type);
    };

    private handleReady: () => void = () => {
        this.navigateToNextIncompleteStep({ isDefault: true });
    };

    private handleNewsletterSubscription: (subscribed: boolean) => void = (subscribed) => {
        this.setState({ isSubscribed: subscribed });
    }

    private handleSignOut: (event: CustomerSignOutEvent) => void = ({ isCartEmpty }) => {
        const { loginUrl, cartUrl, isPriceHiddenFromGuests, isGuestEnabled } = this.props;

        if (isPriceHiddenFromGuests) {
            if (window.top) {
                return (window.top.location.href = cartUrl);
            }
        }

        if (this.embeddedMessenger) {
            this.embeddedMessenger.postSignedOut();
        }

        if (isGuestEnabled) {
            this.setCustomerViewType(CustomerViewType.Guest);
        }

        if (isCartEmpty) {
            this.setState({ isCartEmpty: true });

            if (!isEmbedded()) {
                if (window.top) {
                    return window.top.location.assign(loginUrl);
                }
            }
        }

        this.navigateToStep(CheckoutStepType.Customer);
    };

    private handleShippingNextStep: (isBillingSameAsShipping: boolean) => void = (
        isBillingSameAsShipping,
    ) => {
        this.setState({ isBillingSameAsShipping });

        if (isBillingSameAsShipping) {
            this.navigateToNextIncompleteStep();
        } else {
            this.navigateToStep(CheckoutStepType.Billing);
        }
    };

    private handleShippingSignIn: () => void = () => {
        this.setCustomerViewType(CustomerViewType.Login);
    };

    private handleShippingCreateAccount: () => void = () => {
        this.setCustomerViewType(CustomerViewType.CreateAccount);
    };

    private setCustomerViewType: (viewType: CustomerViewType) => void = (customerViewType) => {
        const { canCreateAccountInCheckout, createAccountUrl } = this.props;

        if (
            customerViewType === CustomerViewType.CreateAccount &&
            (!canCreateAccountInCheckout || isEmbedded())
        ) {
            if (window.top) {
                window.top.location.replace(createAccountUrl);
            }

            return;
        }

        this.navigateToStep(CheckoutStepType.Customer);
        this.setState({ customerViewType });
    };

    private handleBeforeExit: () => void = () => {
        const { analyticsTracker } = this.props;

        analyticsTracker.exitCheckout();
    }
}

export default withExtension(
    withAnalytics(withLanguage(withCheckout(mapToCheckoutProps)(Checkout))),
);
