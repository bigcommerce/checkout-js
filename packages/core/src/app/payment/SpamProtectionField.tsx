import { CheckoutSelectors } from '@bigcommerce/checkout-sdk';
import { noop } from 'lodash';
import React, { Component, MouseEvent, ReactNode } from 'react';

import { CheckoutContextProps, withCheckout } from '../checkout';
import { isErrorWithType } from '../common/error';
import { TranslatedString } from '../locale';
import { LoadingOverlay } from '../ui/loading';

export interface SpamProtectionProps {
    didExceedSpamLimit?: boolean;
    onUnhandledError?(error: Error): void;
}

interface SpamProtectionState {
    shouldShowRetryButton: boolean;
}

interface WithCheckoutSpamProtectionProps {
    isExecutingSpamCheck: boolean;
    executeSpamCheck(): Promise<CheckoutSelectors>;
}

function mapToSpamProtectionProps({
    checkoutService,
    checkoutState,
}: CheckoutContextProps): WithCheckoutSpamProtectionProps {
    return {
        isExecutingSpamCheck: checkoutState.statuses.isExecutingSpamCheck(),
        executeSpamCheck: checkoutService.executeSpamCheck,
    };
}

class SpamProtectionField extends Component<
    SpamProtectionProps & WithCheckoutSpamProtectionProps,
    SpamProtectionState
> {
    state = {
        shouldShowRetryButton: false,
    };

    async componentDidMount() {
        const { didExceedSpamLimit } = this.props;

        if (didExceedSpamLimit) {
            return;
        }

        this.verify();
    }

    render() {
        const { isExecutingSpamCheck } = this.props;

        return (
            <div className="spamProtection-container">
                <LoadingOverlay isLoading={isExecutingSpamCheck}>
                    {this.renderContent()}
                </LoadingOverlay>
            </div>
        );
    }

    private renderContent(): ReactNode {
        const { didExceedSpamLimit } = this.props;
        const { shouldShowRetryButton } = this.state;

        if (!didExceedSpamLimit && !shouldShowRetryButton) {
            return;
        }

        return (
            <div className="spamProtection-panel optimizedCheckout-overlay">
                <a
                    className="spamProtection-panel-message optimizedCheckout-primaryContent"
                    data-test="spam-protection-verify-button"
                    onClick={this.handleRetry}
                >
                    <TranslatedString id="spam_protection.verify_action" />
                </a>
            </div>
        );
    }

    private async verify(): Promise<void> {
        const { executeSpamCheck, onUnhandledError = noop } = this.props;

        try {
            await executeSpamCheck();
        } catch (error) {
            this.setState({ shouldShowRetryButton: true });

            // Notify the parent component if the user experiences a problem other than cancelling the reCaptcha challenge.
            if (
                isErrorWithType(error) &&
                error.type !== 'spam_protection_challenge_not_completed'
            ) {
                onUnhandledError(error);
            }
        }
    }

    private handleRetry: (event: MouseEvent) => void = (event) => {
        event.preventDefault();

        this.verify();
    };
}

export default withCheckout(mapToSpamProtectionProps)(SpamProtectionField);
