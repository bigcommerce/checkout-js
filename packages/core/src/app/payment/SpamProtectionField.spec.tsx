import { CheckoutService, createCheckoutService, StandardError } from '@bigcommerce/checkout-sdk';
import { mount, render } from 'enzyme';
import React, { FunctionComponent } from 'react';

import { CheckoutProvider } from '../checkout';

import SpamProtectionField, { SpamProtectionProps } from './SpamProtectionField';

describe('SpamProtectionField', () => {
    let checkoutService: CheckoutService;
    let SpamProtectionTest: FunctionComponent<SpamProtectionProps>;

    beforeEach(() => {
        checkoutService = createCheckoutService();

        jest.spyOn(checkoutService, 'executeSpamCheck').mockResolvedValue(
            checkoutService.getState(),
        );

        SpamProtectionTest = (props) => (
            <CheckoutProvider checkoutService={checkoutService}>
                <SpamProtectionField {...props} />
            </CheckoutProvider>
        );
    });

    it('renders spam protection field', () => {
        expect(render(<SpamProtectionTest />)).toMatchSnapshot();
    });

    it('notifies parent component if unable to verify', async () => {
        const handleError = jest.fn();
        const error = { type: 'test error' };

        jest.spyOn(checkoutService, 'executeSpamCheck').mockRejectedValue(error);

        const component = mount(<SpamProtectionTest onUnhandledError={handleError} />);

        await new Promise((resolve) => process.nextTick(resolve));
        component.update();

        expect(handleError).toHaveBeenCalledWith(error);
    });

    it('does not notify parent component if unable to verify because of cancellation by user', async () => {
        const handleError = jest.fn();
        const error = { type: 'spam_protection_challenge_not_completed' };

        jest.spyOn(checkoutService, 'executeSpamCheck').mockRejectedValue(error);

        const component = mount(<SpamProtectionTest onUnhandledError={handleError} />);

        await new Promise((resolve) => process.nextTick(resolve));
        component.update();

        expect(handleError).not.toHaveBeenCalledWith(error);
    });

    describe('if have not exceeded limit', () => {
        it('executes spam check on mount', () => {
            mount(<SpamProtectionTest />);

            expect(checkoutService.executeSpamCheck).toHaveBeenCalled();
        });

        it('does not render verify message', () => {
            const component = mount(<SpamProtectionTest />);

            expect(component.exists('[data-test="spam-protection-verify-button"]')).toBeFalsy();
        });

        it('renders verify message if there is error', async () => {
            jest.spyOn(checkoutService, 'executeSpamCheck').mockRejectedValue(
                new Error('Unknown error'),
            );

            const component = mount(<SpamProtectionTest />);

            await new Promise((resolve) => process.nextTick(resolve));
            component.update();

            expect(component.exists('[data-test="spam-protection-verify-button"]')).toBeTruthy();
        });
    });

    describe('if exceeded limit at least once', () => {
        it('does not execute spam check on mount', () => {
            mount(<SpamProtectionTest didExceedSpamLimit />);

            expect(checkoutService.executeSpamCheck).not.toHaveBeenCalled();
        });

        it('executes spam check on click', () => {
            const component = mount(<SpamProtectionTest didExceedSpamLimit />);

            component.find('[data-test="spam-protection-verify-button"]').simulate('click');

            expect(checkoutService.executeSpamCheck).toHaveBeenCalled();
        });

        it('renders verify message', () => {
            const component = mount(<SpamProtectionTest didExceedSpamLimit />);

            expect(component.exists('[data-test="spam-protection-verify-button"]')).toBeTruthy();
        });
    });
});
