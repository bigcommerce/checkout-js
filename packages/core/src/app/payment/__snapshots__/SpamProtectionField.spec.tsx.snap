// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpamProtectionField renders spam protection field 1`] = `
<div
  class="spamProtection-container"
>
  <div
    class="loadingOverlay-container"
  />
</div>
`;
